### PR TITLE
Remove TestPerson #310

### LINKED ISSUE #310: Remove TestPerson
Unless we specifically write unit tests for it to ensure it satisfied all of the contracts of a `ReadOnlyPerson`, having it around is just another potential source of bugs (e.g. violation of LSP) and just a general code smell.

(e.g., for example, I think `name = new Name(); name.getName()` returns null, which a ReadOnlyPerson should never do.)

### PR DESCRIPTION
**Pull request by chao1995**
Fixes #310 

Ready for review.

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@chao1995 submitted v1 for review.

* [v1 1/1] [Remove TestPerson](https://github.com/se-edu/addressbook-level4/pull/335/commits/8dcd2c5f73e8c414eca523b936e174603520a451)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/297506b87fabe6de6baa7901388039ec50199802...8dcd2c5f73e8c414eca523b936e174603520a451))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/335/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/main/java/seedu/address/model/person/Person.java [hunk 1]
```diff
@@ -102,6 +102,19 @@ public void resetData(ReadOnlyPerson replacement) {
         this.setTags(replacement.getTags());
     }
 
+    /**
+     * @return an add command string for adding this person.
+     */
+    public String getAddCommand() {
```
Comment (yamgent) [inline src/main/java/seedu/address/model/person/Person.java]: This method is a helper method for testing purposes, so it seems wrong to put it here in production code.

Comment (chao1995) [inline src/main/java/seedu/address/model/person/Person.java]: True. How about moving it to `AddressBookGuiTest`?

**Comment by CanIHasReview** [comment]
# v2

@chao1995 submitted v2 for review.

* [v2 1/1] [Remove TestPerson](https://github.com/se-edu/addressbook-level4/pull/335/commits/36db4681ec4d713103975b53bc4a61becee1ed7c)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/297506b87fabe6de6baa7901388039ec50199802...36db4681ec4d713103975b53bc4a61becee1ed7c)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/1fe164495cd5c3ce8cfd6d09194f8f2518c8e02a~1...1fe164495cd5c3ce8cfd6d09194f8f2518c8e02a))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/335/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/guitests/AddressBookGuiTest.java [hunk 2]
```diff
@@ -105,6 +106,19 @@ public void cleanup() throws TimeoutException {
     }
 
     /**
+     * @return an add command string for adding this person.
```
Comment (kychua) [inline src/test/java/guitests/AddressBookGuiTest.java]: Should be `Returns an add command ...`

**Comment by CanIHasReview** [comment]
# v3

@chao1995 submitted v3 for review.

* [v3 1/1] [Remove TestPerson](https://github.com/se-edu/addressbook-level4/pull/335/commits/fca41fd0f68bf1ecdbccb37140a0547bc245edd2)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f5efb728b72c193913006d96d4d5591da810af50...fca41fd0f68bf1ecdbccb37140a0547bc245edd2)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/13a93922842c31aff9c5c1ed38bfcc2babb62cd3~1...13a93922842c31aff9c5c1ed38bfcc2babb62cd3))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/335/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v4

@chao1995 submitted v4 for review.

* [v4 1/1] [Remove TestPerson](https://github.com/se-edu/addressbook-level4/pull/335/commits/9d633076bca6600666dd9fd02ad673183ca065e7)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/f5efb728b72c193913006d96d4d5591da810af50...9d633076bca6600666dd9fd02ad673183ca065e7)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/42b0e04ec851544164de4bdee18f2d8e9447815e~1...42b0e04ec851544164de4bdee18f2d8e9447815e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/335/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/guitests/AddressBookGuiTest.java [hunk 3]
```diff
@@ -108,6 +109,19 @@ public void cleanup() throws TimeoutException {
...
 
     /**
+     * Returns an add command string for adding this person.
+     */
+    protected String getAddCommand(Person person) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("add " + person.getName().fullName + " ");
+        sb.append("a/" + person.getAddress().value + " ");
+        sb.append("p/" + person.getPhone().value + " ");
+        sb.append("e/" + person.getEmail().value + " ");
```
Comment (damithc) [inline src/test/java/guitests/AddressBookGuiTest.java]: Any possibility of getting these magic strings from where they are defined?

File: src/test/java/guitests/AddressBookGuiTest.java [hunk 4]
```diff
@@ -108,6 +109,19 @@ public void cleanup() throws TimeoutException {
...
+        person.getTags().asObservableList().stream().forEach(s -> sb.append("t/" + s.tagName + " "));
+        return sb.toString();
+    }
```
Comment (damithc) [inline src/test/java/guitests/AddressBookGuiTest.java]: Also, is there a better place for this method? I'm assuming this functionality will be needed in multiple places.

Comment (chao1995) [inline src/test/java/guitests/AddressBookGuiTest.java]: Yes. There is another `getAddCommand()` method in `LogicManagerTest#TestDataHelper` too. I plan to move them into one place, e.g. under testutil package, `AddCommandBuilder`.

File: src/test/java/guitests/AddressBookGuiTest.java (hunk 3 above)
Comment (chao1995) [inline src/test/java/guitests/AddressBookGuiTest.java]: Will investigate this.

**Comment by chao1995** [comment]
Have some dependency on other PR. Will wait for it to be merged first.

**Comment by CanIHasReview** [comment]
# v5

@chao1995 submitted v5 for review.

* [v5 1/3] [TestPerson: remove methods violating LSP](https://github.com/se-edu/addressbook-level4/pull/335/commits/21c79dc2bdae83dccd4926a824eed9186a107120)
* [v5 2/3] [Extract Prefix class out of ArgumentTokenizer class](https://github.com/se-edu/addressbook-level4/pull/335/commits/53c81cae585e6e47d2d14227f12385d50a1fb4ec)
* [v5 3/3] [TestPerson: remove magic strings in getAddCommand\(\)](https://github.com/se-edu/addressbook-level4/pull/335/commits/380cadcde612b37304fcda7bc2d48706ddcb738b)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...380cadcde612b37304fcda7bc2d48706ddcb738b)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/85732d6ca2513d1507e8c5040de0407702cb8b83~1...85732d6ca2513d1507e8c5040de0407702cb8b83))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/335/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by chao1995** [comment]
@damithc 

> Any possibility of getting these magic strings from where they are defined?

I cherry-picked Pierce's commit from [this PR](https://github.com/se-edu/addressbook-level4/pull/320) (instead of writing my own to reduce merge conflict), which exposes `Prefix`'s `getPrefix` as a public method. Then I can get rid of the magic strings, as in [v5 3/3].

> Also, is there a better place for this method? I'm assuming this functionality will be needed in multiple places.

After trying creating a static method `getAddComand(Person)` elsewhere, I found the best place to put this method is still the `Person` class. So instead of removing `TestPerson`, I decide to extend it from `Person` and put `getAddCommand()` (no args) there.

@yamgent @kychua Please help review the new approach.

**Comment by damithc** [comment]
> After trying creating a static method getAddComand(Person) elsewhere, I found the best place to put this method is still the Person class. So instead of removing TestPerson, I decide to extend it from Person and put getAddCommand() (no args) there.

A person reading the test cases will not know how `TestPerson` differs from `Person`. He has to read the entire `TestPerson` code to confirm it does not differ much. How about we create a separate util class just to host this util method e.g. `PersonUtil#getAddCommand(Person)` which is still cleaner than using a `TestPerson` in tests?

**Comment by chao1995** [comment]
@damithc 

> How about we create a separate util class just to host this util method e.g. PersonUtil#getAddCommand(Person) which is still cleaner than using a TestPerson in tests?

I use this approach in v6. So `TestPerson` class is removed and `PersonUtil` is added to hold `getAddCommand`

**Comment by CanIHasReview** [comment]
# v6

@chao1995 submitted v6 for review.

* [v6 1/3] [Remove TestPerson](https://github.com/se-edu/addressbook-level4/pull/335/commits/6b5e31a414f1a3b1a7ad3f4373c11b110eacf8f2)
* [v6 2/3] [Extract Prefix class out of ArgumentTokenizer class](https://github.com/se-edu/addressbook-level4/pull/335/commits/dc4b1b531d3ce2e7496b8f34f001df12466e5583)
* [v6 3/3] [PersonUtil: remove magic strings in getAddCommand\(\)](https://github.com/se-edu/addressbook-level4/pull/335/commits/fc54745ebf0779d126976a7783e4bd7bca2d79c1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...fc54745ebf0779d126976a7783e4bd7bca2d79c1)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/e3ca3c315e9eefafc3d7734907536687ac3770d5~1...e3ca3c315e9eefafc3d7734907536687ac3770d5))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/335/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by chao1995** [comment]
@damithc 
> How about we create a separate util class just to host this util method e.g. PersonUtil#getAddCommand(Person) which is still cleaner than using a TestPerson in tests?

This approach is used in v6. `TestPerson` is removed and `PersonUtil` is added to hold `getAddCommand()` method.

**Comment by chao1995** [comment]
Some issues with canihasreview. Will start a different PR from v6.

**Pull request closed without merging**